### PR TITLE
Fix elevation profile on closed polyline (fix #58450)

### LINKED ISSUE #58450: Not possible to create a profile from a closed polyline
### What is the bug or the crash?

When you try to create a profile from a closed polyline where the last node coincides with the first node (i.e. it is a closed polyline), the profile is not shown properly.

### Steps to reproduce the issue

See the gif below:
3D-Profile

### Versions

3.38

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dvdkon**
## Description

Currently, the elevation profile code for vector layers (i.e. polylines) goes through each part of the target curve and uses `lineLocatePoint(QgsPoint)` to see how far a point is on the curve. This, however, can't work for closed curves (or in general curves which use a single point multiple times), since there are multiple answers for "how far along the curve is this point?".

This PR fixes the problem by splitting the profile curve into multiple segments which don't self-intersect, and processing each independently. This should fix the issue for all affected polylines while having no effect for previously-working inputs.

Fixes #58450

### CODE REVIEW

File: src/core/vector/qgsvectorlayerprofilegenerator.cpp [hunk 1]
```diff
@@ -721,6 +722,83 @@ bool QgsVectorLayerProfileGenerator::generateProfile( const QgsProfileGeneration
...
+            totalResults->features[it->first].push_back( feature );
+          }
+        }
+      }
+
+      distanceProcessed += mProfileCurve->length();
+      return true;
+    };
+
+    QgsPointSequence allPoints;
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayerprofilegenerator.cpp:771]: I think this logic could be moved to QgsLineString (or similar), as a utility function "splitOverlappingPartsToMultiLineString". I'd also like to see unit tests added covering this.

File: src/core/vector/qgsvectorlayerprofilegenerator.cpp [hunk 2]
```diff
@@ -721,6 +722,83 @@ bool QgsVectorLayerProfileGenerator::generateProfile( const QgsProfileGeneration
...
+    profileLine->points( allPoints );
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayerprofilegenerator.cpp:772]: This is an inefficient approach to iterate through the points in a linestring -- see eg https://github.com/qgis/QGIS/blob/master/src/core/symbology/qgssymbollayerutils.cpp#L4452 . for the fastest approach

File: src/core/vector/qgsvectorlayerprofilegenerator.cpp (hunk 1 above)
Comment (dvdkon) [reply src/core/vector/qgsvectorlayerprofilegenerator.cpp]: Done.

File: src/core/vector/qgsvectorlayerprofilegenerator.cpp (hunk 2 above)
Comment (dvdkon) [reply src/core/vector/qgsvectorlayerprofilegenerator.cpp]: Thanks, I've switched to iterating the data directly.

Is there a reason that there's no iterator of `QgsPoint` on `QgsLineString`?

Comment (nyalldawson) [reply src/core/vector/qgsvectorlayerprofilegenerator.cpp]: > Is there a reason that there's no iterator of QgsPoint on QgsLineString?

They aren't stored natively as QgsPoint in QgsLineString, so any iterator would be constructing a bunch of new objects and would be inefficient as a result. That's why it's recommended to iterate the co-ordinate arrays directly.

On this note it's also MUCH faster to construct QgsLineString from Qvector<double> containers for X/y, rather then the QgsPointSequence approach used here.

Comment (dvdkon) [reply src/core/vector/qgsvectorlayerprofilegenerator.cpp]: > They aren't stored natively as QgsPoint in QgsLineString, so any iterator would be constructing a bunch of new objects and would be inefficient as a result. That's why it's recommended to iterate the co-ordinate arrays directly.

I would hope that the temporary object construction would be optimised by the compiler, but I've never checked. In this specific case it would lead to much nicer code, since I'm constructing `QgsPoint`s anyway.

> On this note it's also MUCH faster to construct QgsLineString from Qvector containers for X/y, rather then the QgsPointSequence approach used here.

I need a single object to put into the `QSet` for duplicate detection, so I can't avoid `QgsPoint`s entirely, but I'll change the method to collect two separate arrays

Comment (nyalldawson) [reply src/core/vector/qgsvectorlayerprofilegenerator.cpp]: > need a single object to put into the QSet for duplicate detection

You possibly want QgsPointXY for that, which will take half the storage space if z/m aren't important

Comment (dvdkon) [reply src/core/vector/qgsvectorlayerprofilegenerator.cpp]: > You possibly want QgsPointXY for that, which will take half the storage space if z/m aren't important

Thanks, I've switched the set to `QgsPointXY`. I've also renamed the method, since a few changes back it started throwing away non-XY coordinates. That's what I want for the elevation profile, but I don't want to surprise people expecting a more general method.

File: src/core/geometry/qgslinestring.cpp [hunk 3]
```diff
@@ -1086,6 +1086,40 @@ std::tuple<std::unique_ptr<QgsCurve>, std::unique_ptr<QgsCurve> > QgsLineString:
...
+  const double *allPointsY = yData();
+  size_t allPointsCount = numPoints();
+  QVector<double> partX;
+  QVector<double> partY;
+  QSet<QgsPointXY> partPointSet;
+
+  QVector<QgsLineString> disjointParts;
+  for ( size_t i = 0; i < allPointsCount; i++ )
+  {
+    QgsPoint point( *allPointsX++, *allPointsY++ );
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.cpp:1101]:
```suggestion
    const QgsPointXY point( *allPointsX++, *allPointsY++ );
```

File: src/core/vector/qgsvectorlayerprofilegenerator.cpp [hunk 4]
```diff
@@ -721,6 +722,70 @@ bool QgsVectorLayerProfileGenerator::generateProfile( const QgsProfileGeneration
   if ( !mProfileCurve || mFeedback->isCanceled() )
     return false;
 
+  auto profileLine = qgsgeometry_cast<QgsLineString *>( mProfileCurve.get() );
+  if ( profileLine )
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayerprofilegenerator.cpp:726]:
```suggestion
  if ( const QgsLineString *profileLine = qgsgeometry_cast< const QgsLineString *>( mProfileCurve.get() ) )
```

File: src/core/vector/qgsvectorlayerprofilegenerator.cpp [hunk 5]
```diff
@@ -721,6 +722,70 @@ bool QgsVectorLayerProfileGenerator::generateProfile( const QgsProfileGeneration
...
+    // point multiple times. We handle this for line strings by splitting them
+    // into multiple parts, each with no repeated points, and computing the
+    // profile for each by itself.
+    QgsCurve *origCurve = mProfileCurve.release();
+    std::unique_ptr< QgsVectorLayerProfileResults > totalResults;
+    double distanceProcessed = 0;
+
+    for ( QgsLineString &part : profileLine->splitToDisjointXYParts() )
+    {
+      mProfileCurve.reset( &part );
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayerprofilegenerator.cpp:738]: This is rather ugly -- we are assigning a unique_ptr a value which it does not have ownership on. I would suggest instead doing something like:

```
   QVector< QgsLineString * > parts = profileLine->splitToDisjointXYParts();
   while ( !parts.isEmpty() )
   {
      mProfileCurve.reset( parts.takeFirst() );
```

(This will require modifying splitToDisjointXYParts to return pointers in the list, which is arguably more flexible anyway)

File: src/core/vector/qgsvectorlayerprofilegenerator.cpp [hunk 6]
```diff
@@ -721,6 +722,70 @@ bool QgsVectorLayerProfileGenerator::generateProfile( const QgsProfileGeneration
...
     return false;
 
+  auto profileLine = qgsgeometry_cast<QgsLineString *>( mProfileCurve.get() );
+  if ( profileLine )
+  {
+    // The profile generation code can't deal with curves that enter a single
+    // point multiple times. We handle this for line strings by splitting them
+    // into multiple parts, each with no repeated points, and computing the
+    // profile for each by itself.
+    QgsCurve *origCurve = mProfileCurve.release();
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayerprofilegenerator.cpp:732]:
```suggestion
    std::unique_ptr< QgsCurve > origCurve = std::move( mProfileCurve );
```

File: tests/src/core/geometry/testqgsgeometry.cpp [hunk 7]
```diff
@@ -751,6 +752,20 @@ void TestQgsGeometry::splitCurve()
...
+  QVector<QgsLineString> onePartParts = onePartLine.splitToDisjointXYParts();
+  QCOMPARE( onePartParts.size(), 1 );
+  QCOMPARE( onePartParts[0].asWkt(), onePartLine.asWkt() );
+
+  QgsLineString triangle( QVector<QgsPoint> {QgsPoint( 0.0, 0.0 ), QgsPoint( 1.0, 0.0 ), QgsPoint( 1.0, 1.0 ), QgsPoint( 0.0, 0.0 )} );
+  QVector<QgsLineString> triangleParts = triangle.splitToDisjointXYParts();
+  QCOMPARE( triangleParts.size(), 2 );
+  QCOMPARE( triangleParts[0].asWkt(), "LineString (0 0, 1 0, 1 1)" );
+  QCOMPARE( triangleParts[1].asWkt(), "LineString (1 1, 0 0)" );
+}
```
Comment (nyalldawson) [inline tests/src/core/geometry/testqgsgeometry.cpp:767]: Can you add a test with an empty line string and a one-point line string too?

File: src/core/vector/qgsvectorlayerprofilegenerator.cpp (hunk 5 above)
Comment (dvdkon) [reply src/core/vector/qgsvectorlayerprofilegenerator.cpp]: It is ugly. I didn't want to have `splitToDisjointXYParts()` return a list of pointers, to avoid allocations.

Returning a list of `unique_ptr` doesn't work with `QVector` (it copies elements on mutation), so I switched to `std::vector`. Sadly now SIP complains that "splitToDisjointXYParts has an unsupported return type", so I've excluded it from the Python bindings for now.

File: tests/src/core/geometry/testqgsgeometry.cpp (hunk 7 above)
Comment (dvdkon) [reply tests/src/core/geometry/testqgsgeometry.cpp:777]: Done.

File: src/core/vector/qgsvectorlayerprofilegenerator.cpp (hunk 5 above)
Comment (nyalldawson) [reply src/core/vector/qgsvectorlayerprofilegenerator.cpp]: >It is ugly. I didn't want to have splitToDisjointXYParts() return a list of pointers, to avoid allocations.

Trust me -- In this case, you'll actually end up with an API which AVOIDS more allocations by returning pointers here. This is because of the mix of geometry classes in QGIS, and the different ways these are used. There's many classes which require QgsGeometry objects (or methods ONLY available from QgsGeometry), so we end up having to create QgsGeometry objects taking ownership of QgsAbstractGeometry in order to call these.  A prime example would be something like QgsFeature -- in order to set the geometry for a feature, we need a QgsGeometry value, and that can ONLY be created with a heap allocated QgsAbstractGeometry.

So by returning values from a method like `splitToDisjointXYParts` we'll end up having to clone those QgsAbstractGeometry values in order to transfer them to a QgsGeometry... which means we're heap allocating anyway PLUS incurring the cost of the clone.

I agree it's not ideal, but we're dealing with a bunch of constraints here which we just have to live with, eg

- Qt API isn't designed around modern c++ memory handling, and FORCES us to use a lot of raw pointers and manual memory management to fit in with their parent/child object model
- The SIP Python binding generator has NO concept of modern c++, and FORCES us to use raw pointers and manual memory management if we want to expose methods to PyQGIS
- We have a stable API constraint forcing us to live with the API decisions from back in the 3.0 era, so we're even constrained by QGIS API itself  :scream_cat: 

I would LOVE to be able to use unique_ptr throughout our API in order to modernise things, but the sad reality of QGIS development is that we just CAN'T do this right now.

> Returning a list of unique_ptr doesn't work with QVector (it copies elements on mutation), so I switched to std::vector. Sadly now SIP complains that "splitToDisjointXYParts has an unsupported return type", so I've excluded it from the Python bindings for now.

As per my comments above, this is something you just need to learn to compromise on, and hold back the vomit :wink:  and expose as an unsafe `QList< pointer *>` return value (along with the `SIP_FACTORY` annotation to indicate that the caller takes ownership of the objects).

Comment (dvdkon) [reply src/core/vector/qgsvectorlayerprofilegenerator.cpp]: Thanks for your rationale on the allocations, that makes sense.

I really don't like returning raw pointers, and lists of them are even worse. But if you think it's the only way for now, I'll change the code. Hopefully one day SIP will support smart pointers and we can refactor at least some of the ugliness away.

File: src/core/vector/qgsvectorlayerprofilegenerator.cpp [hunk 8]
```diff
@@ -721,6 +722,76 @@ bool QgsVectorLayerProfileGenerator::generateProfile( const QgsProfileGeneration
...
+            feature.crossSectionGeometry.translate( distanceProcessed, 0 );
+            totalResults->features[it->first].push_back( feature );
+          }
+        }
+      }
+
+      distanceProcessed += mProfileCurve->length();
+      // We filled mProfileCurve with a reference to part of a vector. We now
+      // need to get it out without free()ing the memory, since we don't own it.
+      Q_UNUSED( mProfileCurve.release() );
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayerprofilegenerator.cpp:782]: Here we DO own the part -- if we don't reset() mProfileCurve here, then we're never cleaning up the parts from disjointParts and we'll leak them

File: src/core/vector/qgsvectorlayerprofilegenerator.cpp (hunk 5 above)
Comment (nyalldawson) [reply src/core/vector/qgsvectorlayerprofilegenerator.cpp]: Thanks for the flexibility! 

> Hopefully one day SIP will support smart pointers and we can refactor at least some of the ugliness away.

It's **possible** we could add support for `std::vector< std::unique_ptr< T > >` return types via some new conversion code in https://github.com/qgis/QGIS/blob/master/python/core/conversions.sip, but AFAIK no-one has attempted that yet...

File: src/core/vector/qgsvectorlayerprofilegenerator.cpp (hunk 8 above)
Comment (dvdkon) [reply src/core/vector/qgsvectorlayerprofilegenerator.cpp]: Gah, I didn't delete this with the rest of the changes, sorry. Thanks for noticing.

Now the parts should be deleted by the moves into `mProfileCurve`.

**Review by nyalldawson — APPROVED** [review]
